[WIP] Refactor the build_system feature to merge it into the fallback feature

Lattay opened the pull request:
This pull request provide an implementation of the ideas discussed in #2.
For short, it aims at providing a homogeneous interface for simplest tasks regardless of the build system.
The interface is mirrored from the existing interface of the fallback system where default targets were defined depending on the filetype.
In effect this PR use the existing fallback code to implement the build system code.

To make clear that targets may be something else that a build target, they are renamed subcommands.

TODO:
- [x] refactor to use only subcommands and not prefix + target
- [x] allow for user customization
- [x] update inline documentation
- [x] update default build systems definitions
- [x] provide `:Build init`
- [x] update doc/build.txt
- [x] update README.md
- [x] test different build systems

Bonus:
- [x] provide `:Build info`
- [x] add internal `:Build help`
- [x] provide a `:Build test`

Lattay committed:
Refactor to provide the same interface wether a build system is found or not

Lattay committed:
Update function code examples, accept globs in the 'file' field of build systems

Lattay commented:
@AlxHnr, I would need some help to make sure the updated build systems are fine because I never used most of them, or not as a developer.
Dune and Make seems ok.
I ran CMake on the Raylib project and it compiled well, but I don't know how sensible is the `run` command.
I think the Autotools config is OK but I only worked on a single project using it, so I don't have too much info on common file architectures used with it.
I tried Cargo on a hello_world type project, it works.
I know nothing about Maven and DUB, and couldn't even write a line of D or Java.

Lattay committed:
provide init help and info subcommands

Lattay committed:
Do not treat init as a special command

Lattay committed:
Store the current build system and provide a BuildRefresh command

Lattay committed:
Update doc

Lattay committed:
Update README

Lattay committed:
Add a test subcommand

Lattay commented:
@AlxHnr I think this is good to go, if you want to review the doc and code.

AlxHnr commented:
I hope to find some time this weekend to look at your work.

> I know nothing about Maven and DUB, and couldn't even write a line of D or Java.

Same for me. I only tried dub on a simple hello-world project five years ago. I don't know if these commands still work today. Only thing I know about Maven is that the build command should use the `compile` target.

`:BuildRefresh` should be moved into its own issue. Multi-build system support has its caveats and should be solved separately. For now, I prefer `:Build` to be as stateless as possible. What if I switch from my current CMake project to a Maven project, but `s:current_build_system` still references CMake? A project like Neovim contains both a CMake file and a normal Makefile. How do we detect which build system is the "main" one?

AlxHnr commented on doc/build.txt:
```diff
@@ -122,18 +160,43 @@ New build systems can be added to |g:build#systems|:
+to the directory of the current file
+                                                                         *%NAME%*
+%NAME% - Shell-escaped filename with its extension.
+                                                                         *%HEAD%*
```
I feel that tagging `%PATH%` is a little excessive and _may_ conflict with other plugins. Its such a short tag.

Lattay replied:
You are right. I think it is important to be able to reach it through help, what do you think about build_%PATH% as a tag ?
Or I simply remove the tags.

AlxHnr replied:
I would prefer to simply remove the tags. In order to search for `:help build_%PATH%` you would already have to know what `%PATH%` is.

AlxHnr commented on autoload/build.vim:
```diff
@@ -203,23 +301,34 @@ endfunction " }}}
+  let l:str = substitute(l:str, '%NAME%',
+        \ escape(shellescape(expand('%:t')), '\'),   'g')
+  let l:str = substitute(l:str, '%HEAD%',
+        \ escape(shellescape(expand('%:t:r')), '\'), 'g')
```
You don't have to wrap at 80 characters. 100 is fine.

Lattay replied:
Ok, I'll fix that

Lattay commented:
> Same for me. I only tried dub on a simple hello-world project five years ago. I don't know if these commands still work today. Only thing I know about Maven is that the build command should use the `compile` target.

Well, at least I should change this to use compile.

> `:BuildRefresh` should be moved into its own issue. Multi-build system support has its caveats and should be solved separately. For now, I prefer `:Build` to be as stateless as possible. 

Yes, I anticipated a bit too much. I introduced a caching Build thinking about your suggestion of `Build chdir`, and so I had to introduce `BuildRefresh` to force update but then I realized that `Build chdir` and other linked features should be a separated PR but I did not reverse the changes introducing `BuildRefresh`.

> What if I switch from my current CMake project to a Maven project, but `s:current_build_system` still references CMake?

It is the reason why I added `BuildRefresh`  it would detect the change when called.

>  A project like Neovim contains both a CMake file and a normal Makefile. How do we detect which build system is the "main" one?

The problem already exists independently of the changes I introduced. Your algo will choose the first encountered.
I thought about adding a `:Build set_build_system` command to manually force a choice, but it should not be in this PR either.

Lattay committed:
Relax line wrapping rule to column 100

Lattay committed:
remove BuildRefresh and Build internal state

Lattay committed:
Prevent conflict by prepending build_ to tags for magics

Lattay commented:
To prevent stupid conflicts in the dune file I propose that it is kept in alphabetical order of the executables names, it should prevent most problems.

Lattay commented:
Should be ok now if you accept the tags `build_%MAGIC%`.

AlxHnr commented:
This plugin was not always stateless. It had a `build#setup()` function, which was called each time a file was created or loaded. After every change to the configuration it was necessary to go trough all open buffers and run `:e` to force re-detection. Thus the feature was removed. I also removed a [chdir feature](https://github.com/AlxHnr/build.vim/commit/8165a69778389a66fdadc22ac462095d6c7d65a3) feature based on this state.

AlxHnr commented on doc/build.txt:
```diff
@@ -122,18 +159,43 @@ New build systems can be added to |g:build#systems|:
+  \       'init' : 'mkdir .foomake_cache',
+  \       'build': 'foomake build',
+  \       'clean': 'foomake wipe',
+  \       'run'  : 'foomake cmd',
```
You should add a placeholder to this example. This shows that commands can do way more and makes people keep reading.

```suggestion
  \       'run'  : 'foomake cmd ./%HEAD%',
```

AlxHnr commented on README.md:
```diff
@@ -2,9 +2,10 @@ This plugin provides commands for using the build system to which the
+* **:Build** - General purpose command for common tasks (build, run, clean...)
 * **:BuildInit** - Initialize/configure a build
 * **:BuildInfo** - Print build informations for the current file
+* **:BuildRefresh** - Force a fresh discovery of build system
```
Not relevant to this PR.

Lattay replied:
Indeed, I'll fix that

AlxHnr commented on README.md:
```diff
@@ -2,9 +2,10 @@ This plugin provides commands for using the build system to which the
-* **:Build** - Build/run a target
+* **:Build** - General purpose command for common tasks (build, run, clean...)
 * **:BuildInit** - Initialize/configure a build
 * **:BuildInfo** - Print build informations for the current file
```
The commands `BuildInit` and `BuildInfo` only exist for backwards compatibility. Leaving them in the README implies they should still be used.

Lattay replied:
Yes, you are right.

AlxHnr commented on README.md:
```diff
@@ -33,10 +34,11 @@ The `:Build` command runs `make` inside the build directory. It takes
+:Build " equivalent to :Build build, build the default target
+:Build build all " build the all target
+:Build test " run the tests
+:Build clean " clean the directory
```
Most of these comments don't say anything.

AlxHnr commented on README.md:
```diff
@@ -122,6 +122,13 @@ Pass custom arguments to the compiler:
 
 For some files which don't belong to a build system, the `run` target will
 be defined:
+It relies on the capacity of the system to directly run those files directly, which imply that you should define a shebang at the begining of the
```
Typo: `imply` -> `implies`

You should move this commend and the python3 block below `:Build run`.

AlxHnr commented on autoload/build.vim:
```diff
@@ -9,38 +9,91 @@ let s:build_systems =
+  \     {
+  \        'do'    : 'cmake --build ./build/ -- -j ' . s:jobs,
+  \        'build' : 'cmake --build ./build/ -- -j ' . s:jobs,
+  \        'clean' : 'rm -r build',
```
```suggestion
  \        'clean' : 'cmake --build ./build/ -- clean',
```

This is what I expect from the command in order to rebuild the project. Otherwise it would _de-initialize_ the project, which is to aggressive.

AlxHnr commented on autoload/build.vim:
```diff
@@ -9,38 +9,91 @@ let s:build_systems =
+  \     'commands' :
+  \     {
+  \        'do'    : 'mvn',
+  \        'build' : 'mvn build',
```
```suggestion
  \        'build' : 'mvn compile',
```

AlxHnr commented on autoload/build.vim:
```diff
@@ -86,7 +139,9 @@ let s:language_cmds =
-  \     'run' : 'ocaml ./%NAME%',
+  \     'build' : 'ocamlc ./%NAME%',
+  \     'clean' : 'rm -f ./%HEAD%',
+  \     'run'   : 'ocaml ./%NAME%',
```
Does the run command care about the output of the build command in this case?

Lattay replied:
No it doesn't. To run the output of the build you would simply use something like `./%HEAD%`.

AlxHnr commented on autoload/build.vim:
```diff
@@ -136,9 +191,33 @@ function! s:log_err(message) " {{{
+" relative to the current working directory
+" Example:
+"   " current working directory is /a/b/c
+"   s:relative_to('/a/b/d/e/f/g', '../d/e')
```
Is this a typo? If its intentional, you should pick a better example.

```suggestion
"   s:relative_to('/a/b/c/d/e/f/g', '../d/e')
```

Lattay replied:
Not a typo, but I have to admit it is not the best example. I'll try to find something better.

AlxHnr commented on autoload/build.vim:
```diff
@@ -136,9 +191,33 @@ function! s:log_err(message) " {{{
+  " expand the destination into a relative path to cwd
+  let l:dest = fnamemodify(l:dest, ':.')
+  " move back to were we started
+  exec 'lcd '.l:initial_wd
```
You don't need `l:initial_wd`.

```suggestion
  lchdir -
```

Please use full command names for consistency. E.g. `exec` -> `execute`, `lcd` -> `lchdir`.

AlxHnr commented on autoload/build.vim:
```diff
@@ -157,25 +236,41 @@ function! s:get_buildsys_item(bs_name, key) " {{{
+" build system. Will return an empty dictionary if nothing is found.
+" If the build_system is not a fallback, both g:build#systems and
+" s:build#systems are considered, the former overwriting the latter.
+" It the build_system is a fallback, both g:build#languages and s:language_cmds
```
```suggestion
" If the build_system is a fallback, both g:build#languages and s:language_cmds
```

AlxHnr commented on autoload/build.vim:
```diff
@@ -157,25 +236,41 @@ function! s:get_buildsys_item(bs_name, key) " {{{
 "
 " Example:
-"   s:gather_fallback_targets('c')
+"   s:gather_commands({'name': 'c', 'fallback': v:true))
```
Why does this function take a dictionary instead of two arguments?

Lattay replied:
It is a `build_system` dictionary as returned by `build#get_current_build_system`. I can change that but for the caller it seems easier to use the full dict and it may be better if `gather_commands` needs more info on the build system.

AlxHnr commented on autoload/build.vim:
```diff
@@ -203,23 +298,31 @@ endfunction " }}}
+function! s:run_command(cmd, extra_args, build_system) " {{{
+  let l:path = a:build_system.fallback? expand('%:p:h') : a:build_system.path
+  let l:command = s:prepare_cmd_for_shell(a:cmd, a:build_system)
+        \ . (empty(a:extra_args)? '' : ' ' . a:extra_args)
```
`extra_args` are not escaped.

AlxHnr replied:
Append them directly to `a:cmd` with a whitespace in between.

AlxHnr commented on autoload/build.vim:
```diff
@@ -262,7 +365,7 @@ endfunction " }}}
   for l:build_name in a:build_system_names
     for l:build_file in split(s:get_buildsys_item(l:build_name, 'file'), ',')
-      if filereadable(a:dir . '/' . l:build_file)
+      if filereadable(a:dir . '/' . l:build_file) || !empty(glob(a:dir . '/' . l:build_file))
```
Vim globbing contains to much magic. Matching full, explicit file names is simpler and gets the job done.

```suggestion
      if filereadable(a:dir . '/' . l:build_file)
```

Lattay replied:
Again it was too much anticipation.
There is the case of Chicken Scheme *chicken-install* which is effectively a build system and use a `name_of_project.egg` file.
Since the change to support glob was so simple it seemed ok to me.
What kind of magic is a problem to you ?

AlxHnr replied:
I see the need to support `*.egg`. But I still have mixed feelings. Glob allows recursive searches via `glob('**')` and even run shell commands:

```vim
glob('`find -iname "*.cpp"`')
```

By default its behaviour is dependent on `wildignore` and `suffixes`.

AlxHnr replied:
What if `a:dir` contains special characters? E.g. the PWD is `/home/user/**`, which is a legal directory name, but gets interpreted by `glob()`. It could be worked around by lchdir + `glob('.' ...)` but thats hacky.

AlxHnr commented on autoload/build.vim:
```diff
@@ -272,19 +375,24 @@ function! s:get_first_build_system_in_dir(dir, build_system_names) " {{{
+" system. If no build system could be found, a fallback build system
+" based on file type is proposed
+" The result is stored on the first call and the research will not be done
+" again.
```
This refers to the removed `:BuildRefresh` commands.

AlxHnr commented on autoload/build.vim:
```diff
@@ -272,19 +375,24 @@ function! s:get_first_build_system_in_dir(dir, build_system_names) " {{{
   if !strlen(l:current_path)
-    return {}
+    let s:current_build_system = {'name': &filetype, 'fallback': v:true, 'path': '.'}
+    return s:current_build_system
```
This variable should've been removed together with `:BuildRefresh`.

```suggestion
    return {'name': &filetype, 'fallback': v:true, 'path': '.'}
```

AlxHnr commented on autoload/build.vim:
```diff
@@ -294,10 +402,12 @@ function! build#get_current_build_system() " {{{
         \ 'path': getcwd(),
+        \ 'fallback': v:false,
         \ }
+      return s:current_build_system
```
This variable should've been removed together with `:BuildRefresh`.

AlxHnr replied:
I see it is still used in many places in this file. Remove those too.

Lattay replied:
Indeed, sorry about that.

AlxHnr commented on autoload/build.vim:
```diff
@@ -328,45 +442,63 @@ endfunction " }}}
+  echo "\n"
+  echo 'Examples:'
+  call s:print_command_examples(l:commands, a:build_system)
+endfunction
```
`:Build info` and `:Build help` should be merged into one function and always do the same thing. This is better for discoverability. Imo `:Build help` should be even removed completely. We have `:help build` for that.

Lattay replied:
I understand for the command but I think it is still nicer to have separated functions internally because it enables a better help message when using a command that is not available.

AlxHnr commented on doc/build.txt:
```diff
@@ -4,7 +4,7 @@ CONTENTS                                                         *build-contents
-    2.1 Build a specific target...........................................|Build|
+    2.1 Common tasks......................................................|Build|
     2.2 Initialize the current project................................|BuildInit|
     2.3 Print build informations for the current file.................|BuildInfo|
```
Remove `:BuildInit` and `:BuildInfo` completely from this help page. The commands are deprecated now and only exist for backwards compatibility.

AlxHnr commented on doc/build.txt:
```diff
@@ -21,47 +21,80 @@ CONTENTS                                                         *build-contents
+
+For example `:Build clean` on a `Make` based project would expand to 
+`make clean` while for a `CMake` based project it would result in
+`rm -r build/`
```
Don't forget to change this CMake command example.

Lattay replied:
What would you expect ? I am not a big user of CMake, but I think most of the time there is no clean target and you don't need to clean unless you want to trash everything and restart from scratch.

AlxHnr replied:
`cmake --build ./build/ -- clean`

AlxHnr replied:
This is consistent with `s:build_systems`. Almost every project has a `clean` target.

AlxHnr commented on doc/build.txt:
```diff
@@ -122,18 +159,43 @@ New build systems can be added to |g:build#systems|:
+
+Every build system must have a name, a file entry and a commands entry.  The
+file entry contains a comma-separated list of filenames or globs (see vim help
+for |glob()|), describing the build files to search for. The commands entry
```
Globs should be removed here too.

AlxHnr commented on doc/build.txt:
```diff
@@ -122,18 +159,43 @@ New build systems can be added to |g:build#systems|:
+Every command will be executed in the directory of the detected build file and can contain
+four different placeholders:
+
+                                                                   *build_%PATH%*
```
Add a tag to this section so they can be referenced in the documentation of `g:build#languages.`.

AlxHnr commented on doc/build.txt:
```diff
@@ -151,12 +213,8 @@ create a dictionary named |g:build#languages|. Here is an example:
-%NAME% - Shell-escaped filename with its extension.
-%HEAD% - Shell-escaped filename without its extension.
+Every command will be executed in the directory of the current file and can
+contain the same placeholders. In this case %RELPATH% always expand to `'.'`.
```
Reference the main placeholder tag here.

AlxHnr reviewed (changes requested):
Im not sure for the defaults of most run targets. Most files don't end up directly in the `build/` directory, so `run: './build/%HEAD%` doesn't make sense. What if a file named `test/foo/bar/mytest.cpp` ends up in `build/some_subdir/test/foo/bar/mytest`?

I don't see how this can be solved in a generic way. Imo the defaults should do the same as `:Build do run`, but can be customized by `g:build#systems`.

Lattay committed:
Update doc/build.txt

Co-authored-by: Alexander Heinrich <AlxHnr@users.noreply.github.com>

Lattay committed:
Update autoload/build.vim

Co-authored-by: Alexander Heinrich <AlxHnr@users.noreply.github.com>

Lattay committed:
Clean up doc and code from references to BuildInfo, BuildRefresh and BuildInit

Lattay committed:
Remove :Build help

Lattay committed:
Stick to target for Build run with CMake

Lattay committed:
More explicit example for relative_to

Lattay committed:
add a placeholder tag

AlxHnr commented:
It seems like you pushed the changes to your master branch, while this merge request still tracks `Lattay:refactor`. You have to either push these commits to this branch or switch this merge requests branch. I've unresolved all discussions and will resolve them myself when you push your changes.

Lattay committed:
Merge branch 'master' into refactor

Lattay commented:
Oh ! Yes sorry about that.

AlxHnr commented on README.md:
```diff
@@ -27,17 +23,17 @@ Open any file in the project and use the following command to initialize
+:Build build all
 :Build test " run the tests
-:Build clean " clean the directory
+:Build clea
```
```suggestion
:Build clean
```

AlxHnr commented on doc/build.txt:
```diff
@@ -26,16 +23,14 @@ the underlying build system (or its absence).
 - `:Build` a general purpose command used to achieve common tasks
 - `:BuildInit` initialize the build system if needed
 - `:BuildInfo` provide informations and help message on the current build
   system
```
You can remove this entire paragraph. `BuildInit` and `BuildInfo` are deprecated.

AlxHnr commented on doc/build.txt:
```diff
@@ -186,6 +156,7 @@ You can omit some of the common commands if they don't make sense in your case.
 Every command will be executed in the directory of the detected build file and can contain
 four different placeholders:
 
+                                                                   *placeholders*
```
```suggestion
                                                             *build_placeholders*
```

AlxHnr reviewed (changes requested):
The last typos.

AlxHnr commented:
The only thing left is the `glob()` discussion, which GitHub seems to hide. I will fix remaining typos myself after merging.

AlxHnr commented:
Are you sure you aren't better off maintaining your own fork? I believe you've got a lot of great ideas and energy, but I can't promise you to have time in the future to collaborate. You will very likely end up with your own fork anyways if you want more features. It took me a week to respond to your pull request and I don't want to waste your time any more. We started with adding a few _dune_ commands and ended up with half a rewrite. I'm not sure I want to maintain all of this.

I don't expect you to do free work. I don't expect anything at all and just don't care. This plugin already does everything I need and upstreaming your ideas won't have any impact since no one uses _build.vim_.

Its free and MIT licensed. Do whatever you want with it.

Lattay committed:
Update README.md

Co-authored-by: Alexander Heinrich <AlxHnr@users.noreply.github.com>

Lattay committed:
Update doc/build.txt

Co-authored-by: Alexander Heinrich <AlxHnr@users.noreply.github.com>

Lattay committed:
Typos and remove glob

Lattay commented:
Hey, I have actually a lot of free time right now because I am between two jobs, but I am sorry I took so much of your time, I can understand that you were not prepared for my contributions, which have been very opinionated as I always am.
Even if no one uses this project it was interesting to me to contribute with a different point of view and I got carried away more than once.
I am ok with fixing more things before you merge if you want, I don't want to be invasive either.
I will probably maintain my own fork indeed for other features, but the features I implemented in this PR seemed profitable to anyone and matched my vision of open-source to submit them to you.

Anyway, thanks for everything :+1:

AlxHnr commented:
I've rebased and merged your changes manually to master. I had to clean up some of the formatting and wording, but it seems to work well. Thanks a lot for your contribution. Some of the issues with `glob()` have been solved by switching to `globpath()`.
